src/foam/dao/AbstractF3FileJournal.js

```diff
@@ -0,0 +1,485 @@
+    },
+    {
+      class: 'Object',
+      name: 'timeStamper',
+      javaType: 'foam.util.FastTimestamper',
+      javaFactory: `return new foam.util.FastTimestamper();`
+    },
+    {
+      class: 'FObjectProperty',
+      of: 'foam.nanos.logger.Logger',
+      name: 'logger',
+      javaFactory: `
+        Logger logger = (Logger) getX().get("logger");
+        if ( logger == null ) {
+          logger = new StdoutLogger();
+        }
+        return new PrefixLogger(new Object[] { "[JDAO]", getFilename() }, logger);
+      `
+    },
+    {
+      class: 'String',
+      name: 'filename',
+      required: true
+    },
+    {
+      class: 'Boolean',
+      name: 'multiLineOutput',
+      value: false
+    },
+    {
+      class: 'Boolean',
+      name: 'createFile',
+      documentation: 'Flag to create file if not present',
+      value: true,
+    },
+    // reader uses a getter because we want a new reader on file replay
+    {
+      class: 'Object',
+      name: 'reader',
+      javaType: 'java.io.BufferedReader',
+      javaGetter: `
+try {
+  InputStream is = getX().get(foam.nanos.fs.Storage.class).getInputStream(getFilename());
+  if ( is == null ) {
+    getLogger().error("File not found - journal: " + getFilename());
+  }
+  return (is == null) ? null : new BufferedReader(new InputStreamReader(is));
+} catch ( Throwable t ) {
+  getLogger().error("Failed to read from journal: " + getFilename(), t);
+  throw new RuntimeException(t);
+}
+      `
+    },
+    // Writer uses a factory because we want to use one writer for the lifetime of this journal object
+    {
+      class: 'Object',
+      name: 'writer',
+      javaType: 'java.io.BufferedWriter',
+      javaFactory: `
+try {
+  OutputStream os = getX().get(foam.nanos.fs.Storage.class).getOutputStream(getFilename());
+  if ( os == null ) {
+    getLogger().error("File not found - journal: " + getFilename());
+  }
+  return (os == null) ? null : new BufferedWriter(new OutputStreamWriter(os));
+} catch ( Throwable t ) {
+  getLogger().error("Failed to read from journal: " + getFilename(), t);
+  throw new RuntimeException(t);
+}
+      `
+    }
+  ],
+
+  methods: [
+    {
+      name: 'put',
+      type: 'FObject',
+      args: [
+        { name: 'x',      type: 'Context' },
+        { name: 'prefix', type: 'String' },
+        { name: 'dao',    type: 'DAO' },
+        { name: 'obj',    type: 'foam.core.FObject' }
+      ],
+      javaCode: `
+        final Object id = obj.getProperty("id");
+
+        getLine().enqueue(new foam.util.concurrent.AbstractAssembly() {
+          FObject old;
+          String  record_ = null;
+
+          public Object[] requestLocks() {
+            return new Object[] { id };
+          }
+
+          public void executeUnderLock() {
+            old = dao.find_(x, id);
+            dao.put_(x, obj);
+          }
+
+          public void executeJob() {}
+
+          public void endJob() {
+            try {
```

Comment:
Here to line 199 we want to move to executeJob() so that this code is done in parallel. The problem with the old Outputter was that it couldn't be used by multiple threads at once, but Formatter doesn't have that problem.